src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```diff
@@ -0,0 +1,212 @@
+                break;
+            case QueueProvider.RabbitMq:
+                ProcessRabbitMq();
+                break;
+        }
+
+        return count;
+    }
+
+    protected override EtlStatsScope CreateScope(EtlRunStats stats)
+    {
+        return new EtlStatsScope(stats);
+    }
+
+    private int ProcessKafka(IEnumerable<QueueWithEvents> items, QueueEtlConfiguration queueEtlConfiguration,
+        DocumentsOperationContext context)
+    {
+        int count = 0;
+
+        using (var jsonSerializer = new BlittableEventBinaryFormatter(context))
+        {
+            var events = new List<KafkaMessageEvent>();
+
+            foreach (QueueWithEvents queueItem in items)
+            {
+                var topicName = queueItem.Name;
+
+                foreach (var insertItem in queueItem.Inserts)
+                {
+                    var eventMessage = new CloudEvent
+                    {
+                        Id = DefaultCloudEventId,
+                        DataContentType = "application/json",
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
+
+                    eventMessage.SetPartitionKey(insertItem.ChangeVector);
+
+                    events.Add(new KafkaMessageEvent
+                    {
+                        Topic = topicName, Message = eventMessage.ToKafkaMessage(ContentMode.Binary, jsonSerializer)
+                    });
+                }
+            }
+
+            if (events.Count == 0) return count;
+
+            if (_kafkaProducer == null)
+            {
+                _kafkaProducer = QueueHelper.CreateKafkaClient(Configuration.Connection, TransactionalId,
+                    Database.ServerStore.Server.Certificate.Certificate);
+                _kafkaProducer.InitTransactions(TimeSpan.FromSeconds(60));
+            }
+
+            void ReportHandler(DeliveryReport<string?, byte[]> report)
+            {
+                if (report.Error.IsError == false)
+                {
+                    count++;
+                }
+                else
+                {
+                    Logger.Info($"Failed to deliver message: {report.Error.Reason}");
+                    _kafkaProducer.AbortTransaction();
+                }
+            }
+
+            try
+            {
+                _kafkaProducer.BeginTransaction();
+
+                foreach (var @event in events)
+                {
+                    _kafkaProducer.Produce(@event.Topic, @event.Message, ReportHandler);
+                }
+
+                _kafkaProducer.CommitTransaction();
+            }
+            catch (Exception ex)
+            {
+                if (ex is KafkaException) _kafkaProducer.AbortTransaction();
```

Comment:
Under what scenario do we want to _not_ abort the transaction here?
What happens if the abort fails? It would mask the actual error, no?

Reply:
Agree, I will fix that.